File: brainfuck.js
// A list of bytes, with each index defaulting to 0, and wrapping to stay
// in the range of 0 - 255.
class ByteList {
  get(i) {
    return this[i] || 0;
  }
  set(i, n) {
    this[i] = n % 256;
  }
}

// Pre-parse a string to find matching pairs of braces.
// Returns an object that contains two sub-objects: `open` and `close`.
// `open` maps the char index of closing brackets to their opening sibling.
// `close` maps the char index of the opening bracket to their closing sibling.
function matchBraces (str, open, close) {
  var result = {
    open: {},
    close: {}
  };
  var opened = [];
  // Loop through and find opening and closing brackets.
  for (var i = 0; i < str.length; i++) {
    if (str.charAt(i) === open) {
      opened.push(i);
    } else if (str.charAt(i) === close) {
      var openInd = opened.pop();
      result.open[i] = openInd;
      result.close[openInd] = i;
    }
  }
  return result;
}

// Brainfuck interpreter - responds with string output, using input as stdin.
function brainfuck(input, stdin) {
  const data = new ByteList(); // Data cells for storage.
  const brackets = matchBraces(input, '[', ']'); // Matching pairs of brackets.

  let index = 0; // Index within data storage.
  let output = ""; // The output string for stdout.
  let stdIndex = 0; // The index within the given stdin.

  // Iterate every character.
  for (var i = 0; i < input.length; i++) {
    switch (input.charAt(i)) {
    case '.':
      output += String.fromCharCode(data.get(index));
      break;
    case '+':
      data.set(index, data.get(index) + 1);
      break;
    case '-':
      data.set(index, data.get(index) - 1);
      break;
    case '>':
      index++;
      break;
    case '<':
      index--;
      break;
    case '[':
      if (!data.get(index)) {
        // Jump to matching closing bracket.
        i = brackets.close[i];
      }
      break;
    case ']':
      i = brackets.open[i] - 1;
      break;
    case ',':
      if (stdIndex < stdin.length)
        data.set(index, stdin.charCodeAt(stdIndex));
      else
        data.set(index, 0);
      stdIndex++;
      break;
    }
  }
  return output;
}

console.log(brainfuck("+[,.]", "This is the brainfuck input text, to be printed!"));

File: brainfuck.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>

using namespace std;

// A list of bytes, defaulting to 0.
class ByteList {
public:
  map<int, char> data;
  unsigned char get(int index);
  void set(int index, unsigned char value);
};

// Get a value in a ByteList.
unsigned char ByteList::get(int i) {
  if (data.find(i) == data.end()) {
    return 0;
  } else {
    return data.at(i);
  }
}

// Set a value in a ByteList.
void ByteList::set(int i, unsigned char v) {
  data[i] = v;
}

// A bidirectional map for parentheses matching.
class ParenSet {
public:
  ParenSet(string in, char opener, char closer);
  int open(int closedAt);
  int close(int openedAt);
  map<int, int> opened;
  map<int, int> closed;
};

// Create the parenthesis set from the given input, using the given
// opening and closing items for parentheses.
ParenSet::ParenSet(string in, char opener, char closer) {
  vector<int> openList;
  for (int i = 0; i < in.size(); i++) {
    char c = in.at(i);
    if (c == opener) {
      openList.push_back(i);
    } else if (c == closer) {
      opened[i] = openList.back();
      closed[openList.back()] = i;
      openList.pop_back();
    }
  }
}

// Get the opening bracket position given a closing bracket position.
int ParenSet::open(int closedAt) {
  if (opened.find(closedAt) == opened.end()) return 0;
  else return opened[closedAt];
}

// Get the closing bracket position given an opening bracket position.
int ParenSet::close(int openedAt) {
  if (closed.find(openedAt) == closed.end()) return 0;
  else return closed[openedAt];
}

// Brainfuck interpreter.
string brainfuck(string code, string in) {
  ParenSet parens(code, '[', ']'); // Parenthesis mapping.
  ByteList memory; // The code's memory stack.

  int inN = 0;     // Index in input text.
  int codeN = 0;   // Index in code text.
  int memN = 0;    // Index in memory.
  string out;

  // Iterate through the brainfuck code.
  while (codeN < code.size()) {
    switch (code.at(codeN)) {

      // Increment / decrement current memory cell.
      case '+':
        memory.set(memN, memory.get(memN) + 1);
        break;
      case '-':
        memory.set(memN, memory.get(memN) - 1);
        break;

      // Move to memory cells to left and right of current position.
      case '>': memN++; break;
      case '<': memN--; break;

      // Print the current character.
      case '.':
        out += memory.get(memN);
        break;

      // Read a character from standard input.
      case ',':
        if (inN < in.size())
          memory.set(memN, in.at(inN));
        else
          memory.set(memN, 0);
        inN++;
        break;

      // Continue or jump conditionally.
      case '[':
        if (!memory.get(memN)) codeN = parens.close(codeN);
        break;
      case ']':
        codeN = parens.open(codeN) - 1;
    }
    codeN++;
  }
  return out;
}

int main () {
  cout << brainfuck("++[,.]", "This is the brainfuck input text, to be printed!") << endl;
}